File: 617-merge-two-binary-trees/merge-two-binary-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def mergeTrees(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> Optional[TreeNode]:
        # #Approach1
        # def helper(x,y):
        #     if not x and not y:
        #         return None
    
        #     if x and y:
        #         curr = TreeNode(x.val + y.val)
        #     elif x:
        #         y = TreeNode()
        #         curr = TreeNode(x.val + y.val)
        #     elif y:
        #         x = TreeNode()
        #         curr = TreeNode(x.val + y.val)

        #     if x.left or y.left:
        #         curr.left = helper(x.left if x.left else TreeNode(), y.left if y.left else TreeNode())
            
        #     if x.right or y.right:
        #         curr.right = helper(x.right if x.right else TreeNode(), y.right if y.right else TreeNode())
        #     return curr

        # return helper(root1,root2)

        if not root1 and not root2:
            return None
        
        value1 = root1.val if root1 else 0

        value2 = root2.val if root2 else 0

        root3 = TreeNode(value1 + value2)

        root3.left = self.mergeTrees(root1.left if root1 else None,root2.left if root2 else None)
        root3.right = self.mergeTrees(root1.right if root1 else None, root2.right if root2 else None)

        return root3

File: 617-merge-two-binary-trees/merge-two-binary-trees.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func mergeTrees(_ root1: TreeNode?, _ root2: TreeNode?) -> TreeNode? {
        return self.recursive(left_child : root1, right_child : root2)
    }

    func recursive(left_child root1: TreeNode?, right_child root2: TreeNode?) -> TreeNode?{
        if let left: TreeNode = root1, let right: TreeNode = root2{
            let curr_value = left.val + right.val;
            print(curr_value)
            var left_child: TreeNode? = self.recursive(left_child : left.left, right_child : right.left);
            var right_child: TreeNode? = self.recursive(left_child : left.right, right_child : right.right);
            var curr_node: TreeNode? = TreeNode(curr_value);
            if let left_child = left_child{
                curr_node?.left = left_child;
            }else{
                curr_node?.left = nil;
            }
            if let right_child = right_child{
                curr_node?.right = right_child;
            }else{
                curr_node?.right = nil;
            }
            
            return curr_node;





        }else if let left: TreeNode = root1{
            let curr_value = left.val;
            print(curr_value)
            var left_child: TreeNode? = self.recursive(left_child : left.left, right_child : nil);
            var right_child: TreeNode? = self.recursive(left_child : left.right, right_child : nil);
            var curr_node: TreeNode? = TreeNode(curr_value)
            if let left_child = left_child{
                curr_node?.left = left_child;
            }else{
                curr_node?.left = nil;
            }
            if let right_child = right_child{
                curr_node?.right = right_child;
            }else{
                curr_node?.right = nil;
            }
            return curr_node;




        }else if let right: TreeNode = root2{
            let curr_value = right.val;
            print(curr_value)
            var left_child: TreeNode? = self.recursive(left_child : nil, right_child : right.left);
            var right_child: TreeNode? = self.recursive(left_child : nil, right_child : right.right);
            var curr_node: TreeNode? = TreeNode(curr_value)
            if let left_child = left_child{
                curr_node?.left = left_child;
            }else{
                curr_node?.left = nil;
            }
            if let right_child = right_child{
                curr_node?.right = right_child;
            }else{
                curr_node?.right = nil;
            }
            return curr_node;
        }else{
            return nil
        }
    }
}